tests/functional/oauth_webchannel_keys.js

```diff
@@ -0,0 +1,510 @@
+        '123done': true
+      });
+    },
+
+    'signup, verify same browser with original tab open': function () {
+      var self = this;
+
+      return openFxaFromRp(self, 'signup')
+        .execute(FunctionalHelpers.listenForWebChannelMessage)
+
+        .then(function () {
+          return FunctionalHelpers.fillOutSignUp(self, email, PASSWORD, OLD_ENOUGH_YEAR);
+        })
+
+        .findByCssSelector('#fxa-confirm-header')
+        .end()
+
+        .then(function () {
+          return FunctionalHelpers.openVerificationLinkSameBrowser(
+                      self, email, 0);
+        })
+        .switchToWindow('newwindow')
+        // wait for the verified window in the new tab
+        .findById('fxa-sign-up-complete-header')
+        // the verification tab will almost always be the one to complete the flow.
+        // XXX TODO: can we not fail the test if the other tab wins?
+        // XXX TODO: or, ensure that the original tab always wins if it's open?
+        .then(testIsBrowserNotifiedOfLogin(self, { shouldCloseTab: false }))
+        .end()
+
+        .closeCurrentWindow()
+        // switch to the original window
+        .switchToWindow('')
+
+        .findById('fxa-sign-up-complete-header')
+        .end();
+    },
+
+    'signup, verify same browser with original tab closed': function () {
+      var self = this;
+
+      return openFxaFromRp(self, 'signup')
+
+        .then(function () {
+          return FunctionalHelpers.fillOutSignUp(self, email, PASSWORD, OLD_ENOUGH_YEAR);
+        })
+
+        .findByCssSelector('#fxa-confirm-header')
+        .end()
+
+        .then(FunctionalHelpers.openExternalSite(self))
+
+        .then(function () {
+          return FunctionalHelpers.openVerificationLinkSameBrowser(self, email, 0);
+        })
+
+        .switchToWindow('newwindow')
+        .execute(FunctionalHelpers.listenForWebChannelMessage)
+
+        .then(testIsBrowserNotifiedOfLogin(self, { shouldCloseTab: false }))
+
+        .findById('fxa-sign-up-complete-header')
+        .end()
+
+        .closeCurrentWindow()
+        // switch to the original window
+        .switchToWindow('');
+    },
+
+    'signup, verify same browser, replace original tab': function () {
+      var self = this;
+
+      return openFxaFromRp(self, 'signup')
+
+        .then(function () {
+          return FunctionalHelpers.fillOutSignUp(self, email, PASSWORD, OLD_ENOUGH_YEAR);
+        })
+
+        .findByCssSelector('#fxa-confirm-header')
+        .end()
+
+        .then(function () {
+          return FunctionalHelpers.getVerificationLink(email, 0);
+        })
+        .then(function (verificationLink) {
+          return self.get('remote').get(require.toUrl(verificationLink))
+            .execute(FunctionalHelpers.listenForWebChannelMessage);
+        })
+
+        .then(testIsBrowserNotifiedOfLogin(self, { shouldCloseTab: false }))
+
+        .findById('fxa-sign-up-complete-header')
+        .end();
+    },
```

Comment:
> The 2 tests above: Is original tab closed vs replace original tab a bit redundant? 

For this case, possibly, for other cases they are not. For example, with the redirect flow, if the user replaces the tab they signed up in with the verification tab, then the user should be redirected back to the relier automatically. If they verify in a 2nd tab and the original tab has closed, then no automatic redirection.